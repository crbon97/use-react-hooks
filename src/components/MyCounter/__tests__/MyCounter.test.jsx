import MyCounterUseState, {
  CounterProvider,
  MyCounterReducer,
  MyCounterWithContext
} from "..";
import { cleanup, fireEvent, getByTestId, render } from "react-testing-library";

import React from "react";
import renderer from "react-test-renderer";

describe("Test MyCounter(s) component with `react-test-renderer`", () => {
  test("MyCounterReducer should render without error", () => {
    const myCounter = renderer.create(<MyCounterReducer />);
    let tree = myCounter.toJSON();
    expect(tree).toMatchSnapshot();
  });
  test("MyCounterUseState should render without error", () => {
    const myCounter = renderer.create(<MyCounterUseState />);
    let tree = myCounter.toJSON();
    expect(tree).toMatchSnapshot();
  });
  test("MyCounterWithContext should render without error", () => {
    const myCounter = renderer.create(
      <CounterProvider>
        <MyCounterWithContext />
      </CounterProvider>
    );
    let tree = myCounter.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

const testCounterBehavior = counterContainer => {
  const buttonDecrease = getByTestId(counterContainer, "button-decrease");
  const buttonIncrease = getByTestId(counterContainer, "button-increase");
  const buttonReset = getByTestId(counterContainer, "button-reset");
  const counterValue = getByTestId(counterContainer, "counter-value");

  // initial props
  expect(buttonDecrease).toBeTruthy();
  expect(counterValue).toBeTruthy();
  expect(buttonIncrease).toBeTruthy();
  expect(buttonReset).toBeTruthy();
  expect(counterValue.textContent).toEqual("0");

  // update prop with action `decrease`
  fireEvent.click(buttonDecrease);
  expect(counterValue.textContent).toEqual("-1");

  // update prop with action `reset`
  fireEvent.click(buttonReset);
  expect(counterValue.textContent).toEqual("0");

  // update prop with action `increase`
  fireEvent.click(buttonIncrease);
  expect(counterValue.textContent).toEqual("1");
};

describe("Test MyCounter(s) component with `react-testing-library`", () => {
  afterEach(cleanup);
  test("MyCounterReducer should render without error", () => {
    const { container } = render(<MyCounterReducer />);
    testCounterBehavior(container);
  });

  test("MyCounterUseState should render without error", () => {
    const { container } = render(<MyCounterUseState />);
    testCounterBehavior(container);
  });
  test("MyCounterWithContext should render without error", () => {
    const { container } = render(
      <CounterProvider>
        <MyCounterWithContext />
      </CounterProvider>
    );
    testCounterBehavior(container);
  });
});
