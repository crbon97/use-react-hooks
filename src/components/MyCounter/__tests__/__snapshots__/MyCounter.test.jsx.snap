// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test MyCounter(s) component with \`react-test-renderer\` MyCounterReducer should render without error 1`] = `
<div
  className="button-group"
>
  <button
    className="button button--left"
    data-testid="button-decrease"
    onClick={[Function]}
  >
    -
  </button>
  <div
    className="value"
    data-testid="counter-value"
  >
    0
  </div>
  <button
    className="button button--right"
    data-testid="button-increase"
    onClick={[Function]}
  >
    +
  </button>
  <button
    className="button button--full-width"
    data-testid="button-reset"
    onClick={[Function]}
  >
    reset
  </button>
</div>
`;

exports[`Test MyCounter(s) component with \`react-test-renderer\` MyCounterUseState should render without error 1`] = `
<div
  className="button-group"
>
  <button
    className="button button--left"
    data-testid="button-decrease"
    onClick={[Function]}
  >
    -
  </button>
  <div
    className="value"
    data-testid="counter-value"
  >
    0
  </div>
  <button
    className="button button--right"
    data-testid="button-increase"
    onClick={[Function]}
  >
    +
  </button>
  <button
    className="button button--full-width"
    data-testid="button-reset"
    onClick={[Function]}
  >
    reset
  </button>
</div>
`;

exports[`Test MyCounter(s) component with \`react-test-renderer\` MyCounterWithContext should render without error 1`] = `
<div
  className="button-group"
>
  <button
    className="button button--left"
    data-testid="button-decrease"
    onClick={[Function]}
  >
    -
  </button>
  <div
    className="value"
    data-testid="counter-value"
  >
    0
  </div>
  <button
    className="button button--right"
    data-testid="button-increase"
    onClick={[Function]}
  >
    +
  </button>
  <button
    className="button button--full-width"
    data-testid="button-reset"
    onClick={[Function]}
  >
    reset
  </button>
</div>
`;
